### PR TITLE
codegen: more capable rewrite to expm1 & cosm1 etc.

### PR DESCRIPTION
**Pull request by bjodah**
#### References to other Issues or PRs
This PR also adds a regression test for gh-21531.

#### Brief description of what is fixed or changed
I reworked the logic behind rewriting of e.g. `exp(x)-1` and `cos(x)-1` to `expm1(x)` and `cosm1(x)` respectively. I was able to remove the `@XFAIL` decorators for some tests. Hopefully this is an improvement for all cases.

I also made the underlying logic public in the form of a class `FuncMinusOneOptim`.

#### Other comments
cc @lcontento, just FYI. I'm also interested to hear of your experiences of using these optimizations, if you want/can share that is.

Perhaps FuncMinusOneOptim is not flexible enough and we should have a class which generalizes to functions subtracting a number other than one. I could imagine e.g. `arccos(x)-pi/2` being a canddiate. However, I have not been able to find any references to such a function when I did a quick internet search. (I wonder what it would be named, `acosmhalfpi`?)
#### Release Notes

* codegen
  * More capable rewriting of terms such as `2*exp(x) + 3*exp(y) - 5` into `2*expm1(x) + 3*expm1(y)`.

### CODE REVIEW

**Comment by bjodah** [comment]
OK, now this PR does more than just add a regression test. I updated the description accordingly. @lcontento would you be interested in reviewing my suggested changes?

**Comment by lcontento** [comment]
> @lcontento would you be interested in reviewing my suggested changes?

I have not that much experience in optimizing expressions for floating point computation and as far as I know it's far from being a solved problem.
There are some automated approaches to do it but they usually result in much less readable expressions.
So I think having some heuristics that deal with the common cases is probably more than enough and the new heuristic you propose seems to be sound and an improvement over the old one. Code looks also good as far as I can see.

However, there is at least one case where the old code does a better job.
```python
from sympy.codegen.rewriting import optimize, expm1_opt
from sympy import sympify
expr = sympify("(1 + exp(x))^2 - 4")
optimize(expr, [expm1_opt])
```
I believe this expression should display the same numerical problems near `x = 0`.
Your branch leaves it unchanged, while the original rewriting logic simplifies it to `(exp(x) + 3)*expm1(x)`, which should be more accurate.
This is due to the fact that the original code always tries factoring before substitution.
Obviously no heuristic can deal with all cases but it may be worth and not too difficult to add back the factorization step.

**Comment by bjodah** [comment]
Yes, heuristics have their limits. I was using this myself for a while (and it did help in some cases), but since then, and in that project, we have moved to polynomials (splines) so right now I'm not using this.

Excellent, exactly the kind of feedback I was hoping for. I'll add your example to the test suite, and will try to rework my patch here (as soon as I get some spare time).

File: sympy/codegen/rewriting.py [hunk 1]
```diff
@@ -175,42 +175,99 @@ def optimize(expr, optimizations):
...
-        factored = protected.factor()
-        new_old = {v: self.func(k) for k, v in old_new.items()}
-        return factored.replace(_d - 1, lambda d: self.func_m_1(new_old[d].args[0])).xreplace(new_old)
-
-    def __call__(self, e):
-        numbers, non_num = sift(e.args, lambda arg: arg.is_number, binary=True)
-        non_num_func, non_num_other = sift(non_num, lambda arg: arg.has(self.func),
-            binary=True)
+    def _group_Add_terms(self, add):
+        numbers, non_num = sift(add.args, lambda arg: arg.is_number, binary=True)
```
Comment (smichr) [inline sympy/codegen/rewriting.py:226]: I wonder if you should further sift numbers into Rational and other:
```python
>>> optimize(2*exp(x)-2,[expm1_opt]).subs(x,1e-11)
2.00000000001000e-11
>>> (2*exp(x)-2).subs(x,1e-11)
2.00000016548074e-11

>>> optimize(2*exp(x)-pi,[expm1_opt])
2*expm1(x) - pi + 2
>>> _.subs(x,1e-11)
2.00000000002 - pi
>>> (2*exp(x)-pi).subs(x,1e-11)
2.00000000002 - pi
```

Comment (bjodah) [reply sympy/codegen/rewriting.py:226]: Ah, yes, not very beautiful, but I am not sure how `is_rational` is central to avoiding cancellation? For example, if I carefully construct a fraction of pi close to two:
```python
In [31]: about_two = 2**(58/S(117))*3**(97/S(117))*5**(4/S(39))*7**(92/S(117))/S(30)*pi

In [32]: xval = S(10)**-11

In [33]: ref = (2*exp(x) - about_two).subs(x, xval).n(17); print(ref)
2.0418173913673214e-11

In [34]: no_opt = (2*exp(x) - about_two).subs(x, xval.evalf()).evalf()

In [35]: with_opt = optimize(2*exp(x) - about_two, [expm1_opt]).n(25).subs(x, 1e-11).evalf()

In [36]: with_opt - ref, no_opt - ref
Out[36]: (1.1536301877952077e-26, 1.6547074214222335e-18)
```
somewhat orthogonal to what we are discussing: note how I needed to insert `.n(25)` there, `evalf` didn't cut it. Hopefully our (code) printers compute such sums to the full precision of the targeted data type... (but this is another issue of course).

Comment (smichr) [reply sympy/codegen/rewriting.py:226]: What the example I gave might be saying is that there is no benefit to rewriting unless you can absorb an integer; if the number is not rational then it looks like there is no benefit. 

As far as the examples go...I'm nervous about the multiple evalfs that you use to demonstrate the difference. I need a more direct demonstration that the optimization is more precise.

Comment (bjodah) [reply sympy/codegen/rewriting.py:226]: Right, I wanted to make sure myself, so I did spend some time adding a test which does a full roundtrip to (generated) C code and back. I just added those commits here.

File: sympy/codegen/tests/test_rewriting.py [hunk 2]
```diff
@@ -297,3 +349,63 @@ def check(d):
...
+    return func_rewritten(x)
+'''
+    with tempfile.TemporaryDirectory() as folder:
+        mod, info = compile_link_import_strings(
+            [('func.c', func_c), ('_func.pyx', func_pyx)],
+            build_dir=folder, compile_kwargs=dict(std='c99')
+        )
+        err_rewritten = abs(mod.py_rewritten(1e-11) - ref)
+        err_unchanged = abs(mod.py_unchanged(1e-11) - ref)
+        assert 1e-27 < err_rewritten < 1e-25  # highly accurate.
```
Comment (bjodah) [inline sympy/codegen/tests/test_rewriting.py:404]: And here is where I show that the accuracy is improved by several orders of magnitude, even though we are dealing with an irrational number.

File: sympy/printing/tests/test_str.py [hunk 3]
```diff
@@ -252,6 +252,8 @@ def test_Mul():
     # For issue 14160
     assert str(Mul(-2, x, Pow(Mul(y,y,evaluate=False), -1, evaluate=False),
                                                 evaluate=False)) == '-2*x/(y*y)'
+    # issue 21537
+    assert str(Mul(x, Pow(1/y, -1, evaluate=False), evaluate=False)) == 'x/(1/y)'
```
Comment (bjodah) [inline sympy/printing/tests/test_str.py:257]: Umm, I don't think I've actively added this line. Weird, I have merged with master...

File: sympy/codegen/tests/test_rewriting.py [hunk 4]
```diff
@@ -297,3 +349,63 @@ def check(d):
...
+
+double func_unchanged(double x) {
+    return %(unchanged)s;
+}
+double func_rewritten(double x) {
+    return %(rewritten)s;
+}
+''' % dict(unchanged=ccode(unchanged.n(NUMBER_OF_DIGITS)),
+           rewritten=ccode(rewritten.n(NUMBER_OF_DIGITS)))
+
```
Comment (bjodah) [inline sympy/codegen/tests/test_rewriting.py:387]: And this is what `func_c` looks like once rendered:
```C
#include <math.h>

double func_unchanged(double x) {
    return 2.0*exp(x) - 1.9999999999995818;
}
double func_rewritten(double x) {
    return 2.0*expm1(x) + 4.1817391357322551e-13;
}
```

**Comment by smichr** [comment]
OK, thanks @bjodah

**Comment by bjodah** [comment]
Thanks for the review @smichr

**Comment by bjodah** [comment]
And thank you @lcontento !

**Pull request merged**